lib/handlers/connect.js

```diff
@@ -33,33 +40,44 @@ var errorMessages = [
 ]
 
 function handleConnect (client, packet, done) {
-  client.connected = true
-  client.clean = packet.clean
-
-  if (!packet.clientId && packet.protocolVersion === 3) {
-    client.emit('error', new Error('Empty clientIds are supported only on MQTT 3.1.1'))
-    return done()
-  }
-
-  client.id = packet.clientId || uuid.v4()
-  client._will = packet.will
-
   clearTimeout(client._connectTimer)
   client._connectTimer = null
 
-  if (packet.keepalive > 0) {
-    client._keepaliveInterval = (packet.keepalive * 1500) + 1
-    client._keepaliveTimer = retimer(function keepaliveTimeout () {
-      client.broker.emit('keepaliveTimeout', client)
-      client.emit('error', new Error('keep alive timeout'))
-    }, client._keepaliveInterval)
+  if (client.broker.preConnect(client) === false) {
+    return client.conn.destroy()
```

Comment:
please do this in two statements, as having multiple returns types make the V8 optimizer sweat a bit more.